library/psa_crypto_aead.c

```diff
@@ -356,12 +328,455 @@ psa_status_t mbedtls_psa_aead_decrypt(
+                                                        input_length,
+                                                        input,
+                                                        output ) );
+    }
+    else
+#endif /* MBEDTLS_PSA_BUILTIN_ALG_GCM */
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_CHACHA20_POLY1305)
+    if( operation->alg == PSA_ALG_CHACHA20_POLY1305 )
+    {
+        status = mbedtls_to_psa_error(
+           mbedtls_chachapoly_update( &operation->ctx.chachapoly,
+                                      input_length,
+                                      input,
+                                      output ) );
+    }
+    else
+#endif /* MBEDTLS_PSA_BUILTIN_ALG_CHACHA20_POLY1305 */
+    {
+        (void) input;
+        (void) input_length;
+
+        return ( PSA_ERROR_NOT_SUPPORTED );
+    }
+
+    if( status == PSA_SUCCESS )
+    {
+        *output_length = update_output_length;
+        operation->body_started = 1;
+    }
+
+    return( status );
+}
+
+/* Common checks for both mbedtls_psa_aead_finish() and
+   mbedtls_psa_aead_verify() */
+static psa_status_t mbedtls_psa_aead_finish_checks(
+    mbedtls_psa_aead_operation_t *operation,
+    size_t output_size,
+    size_t tag_size )
+{
+    size_t finish_output_size;
+
+    if( tag_size < operation->tag_length )
+        return ( PSA_ERROR_BUFFER_TOO_SMALL );
+
+    finish_output_size = operation->is_encrypt ?
+        PSA_AEAD_FINISH_OUTPUT_SIZE( operation->key_type, operation->alg ) :
+        PSA_AEAD_VERIFY_OUTPUT_SIZE( operation->key_type, operation->alg );
+
+    if( output_size < finish_output_size )
+        return ( PSA_ERROR_BUFFER_TOO_SMALL );
+
+    return ( PSA_SUCCESS );
+}
+
+/* Finish encrypting a message in a multipart AEAD operation. */
+psa_status_t mbedtls_psa_aead_finish(
+    mbedtls_psa_aead_operation_t *operation,
+    uint8_t *ciphertext,
+    size_t ciphertext_size,
+    size_t *ciphertext_length,
+    uint8_t *tag,
+    size_t tag_size,
+    size_t *tag_length )
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+    size_t finish_output_size = 0;
+
+    status = mbedtls_psa_aead_finish_checks( operation, ciphertext_size,
+                                             tag_size );
+
+    if( status != PSA_SUCCESS )
+        return status;
+
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_GCM)
+    if( operation->alg == PSA_ALG_GCM )
+        /* We will need to do final GCM pass in here when multipart is done. */
+        status =  mbedtls_to_psa_error( mbedtls_gcm_finish( &operation->ctx.gcm,
+                                                            tag,
+                                                            tag_size ) );
+    else
+#endif /* MBEDTLS_PSA_BUILTIN_ALG_GCM */
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_CHACHA20_POLY1305)
+    if( operation->alg == PSA_ALG_CHACHA20_POLY1305 )
+        status = mbedtls_to_psa_error(
+           mbedtls_chachapoly_finish( &operation->ctx.chachapoly,
+                                      tag ) );
```

Comment:
There is a check on tag_size above, and operation->tag_length for and ChaChaPoly alg is always 16. Is this not enough?

Reply:
Ah, got it, the code is correct here because `tag_size` is guaranteed to be `>= operation->tag_length`, and `psa_aead_setup` enforces that `operation->tag_length` is 16 for Chachapoly. Please at least add a comment here to explain, and a comment in `psa_aead_setup` to say that if shorter tags start to be supported then `mbedtls_psa_aead_finish` needs to be updated as well.

I'd still prefer to have the check right here. `mbedtls_chachapoly_finish` needs a 16-byte buffer so check `tag_size` against 16, not against some value which is guaranteed to be 16 through a somewhat complex reasoning.